### ISSUE TITLE
libssl versioning

### ISSUE THREAD

**Issue by TurkeyMan** [premake/premake-core#528]
I just built head, and I've noticed a new dependency on libssl. When I build myself a premake binary (linux), it creates links to the particular version of libssl I have on my machine.

If I then take that binary and run it on another linux machine (the build machine), it complains the libssl version is wrong:
```
bin/premake/premake5: error while loading shared libraries: libssl.so.1.0.0: cannot open shared object file: No such file or directory
```

Listing the libssl version(/s) present on the build machine:
```
	libssl3.so -> libssl3.so
	libssl.so.10 -> libssl.so.1.0.1e
	libssl.so.6 -> libssl.so.0.9.8e
	libssl3.so -> libssl3.so
	libssl.so.10 -> libssl.so.1.0.1e
```

It doesn't seem right that it linked to libssl.so.1.0.0 precisely... usually it links to something like libssl.so.1 (which is symlink to minor/patch revisions). I think this is a build configuration problem. I haven't looked into it in detail.

**Comment by tvandijck** [comment]
I made a version locally here that uses mbedtls instead, and statically links it, since it builds the library from source in the project... so much more stable and way easier to move around different linux distributions...

that said... the way premake5 sets up the link against ssl is like this:
```lua
		filter "system:linux"
			if not _OPTIONS["no-curl"] and os.findlib("ssl") then
				links       { "ssl", "crypto" }
			end
```

so no 'specific' version specified, and that is really just what your gcc makes of it...
even the generated makefile just specifies "-lssl -lcrypto"

It is true that moving things between different linux distributions is a pain in the <beep> when linking against dynamic modules... locally we even run into issues with "libc.so" between Centos 5 and Centos 7... I don't understand why Linux can't just have binary compatibility. I can make an executable on Windows 95 that will still run on Windows 10.

**Comment by TurkeyMan** [comment]
> so no 'specific' version specified, and that is really just what your gcc makes of it...
even the generated makefile just specifies "-lssl -lcrypto"

Yeah, I saw that, but apparently that decision to leave it to the linker is allowing the linker to do the wrong thing.
I changed it to: `links       { ":libssl.a", ":libcrypto.a" }`, and that worked. I didn't try explicitly linking to `libssl.so.1`, but I suspect that would be the superior choice; my premake binary was huge after static linking ssl and crypto.

> I don't understand why Linux can't just have binary compatibility. I can make an executable on Windows 95 that will still run on Windows 10.

This is common linux problem with closed-source binaries. One way to mitigate is to link against less-specific libs, ie, in our case `libssl.so.1` instead of `libssl.so.1.0.0`. That will allow a much wider range of library versions to be compatible with the app. The alternative is to bundle the libs you use with the exe, like you already have to do on windows.

Windows also protects against this problem with system libraries with their WindowsSxS (side by side) 'solution', which is basically keeping parallel copies of the suite of libs for every revision of windows ever released all together. This is also one of the major reasons my windows install is 120gb, and linux is 2gb, and also why a live Windows install seems to just grow larger and larger with age. I don't think their solution is a good solution ;) .. Linux has an advantage, since most software is OSS, they just rebuild everything against new releases. It just shifts an annoying responsibility to us closed-source devs to either static-link, carefully test and support various dynamic versions, or bundle libs with our binaries.
